### DISCUSSION TITLE
CMYK image after compositeImage appears washed out

### DISCUSSION THREAD

**Discussion by neekfenwick** [ImageMagick/ImageMagick#4211] [Help]
This may be similar to #4208 since the colours also appear washed out for me but I think the root cause is different.

Production Platform: Centos 7.3 - ImageMagick6-6.9.12.21-1.el7.remi.x86_64
Dev Platform: Fedora 34 - ImageMagick-6.9.11.27-3.fc34.x86_64

Both give identical results as far as I can tell.  See attached PHP script and input file "3 Glasses Of Wine.jpg".
![source_and_output_screenshot](https://user-images.githubusercontent.com/266415/133029316-b516cf41-086c-48fe-8e24-200af26f32dc.jpg)
Expected result: cropped output file should have a white background and circular section from original image.
Actual result: cropped output has correct circle but colours are washed out compared to original, see screenshot.

We have large JPG images in CMYK colorspace provided by our graphic artist for printing on industrial printers.  Our website needs to process these JPG files, in this case it wants to crop a circular image from a square master image.

I borrowed code from somewhere that prepares a white circle as a mask and performs a `compositeImage` operation and then `mergeImageLayers` to get the desired result, however the result image has washed out colours.  Simply scaling the image and saving it produces a good looking file.

If I convert the original file to SRGB colorspace before the operation, it produces good looking output, see line 49 of the attached script.  I know that sometimes the viewer being used may produce poor output due to CMYK support but the source image views OK using my Gnome viewer so I think the output file should appear the same.

In case the `$circle` image used for the circular mask is somehow the wrong colour profile, I've tried using something like `$circle->profileImage('icc', $img->getImageProfile('icc'))` to force it to have the same profile as the main image but this has no effect.

Is this something I'm doing wrong, or a bug somewhere?

Would it help if I try to reproduce this using the 'convert' (or similar) command line, rather than working through the PHP imagick bindings?

[crop_image_circle_standalone3.php.txt](https://github.com/ImageMagick/ImageMagick/files/7151653/crop_image_circle_standalone3.php.txt)

![3 Glasses Of Wine-500](https://user-images.githubusercontent.com/266415/133028170-abdd5a30-cba1-404d-ad22-5e85527fb7a1.jpg)

**Comment by snibgo** [comment]
The colours appear washed-out, but are they really?

Colours against a black background appear more intense, more saturated, more colorful than the same colours against a gray (or white) background. I forget the name for this phenomenon.

[EDIT: The effect is "simultaneous contrast". See https://en.wikipedia.org/wiki/Contrast_effect#Simultaneous_contrast.]

Downloading your screenshot of three images to a pixel viewer, it seems that corresponding pixels have identical values, hence identical colours.

**Comment by snibgo** [comment]
A couple of comments:

For operations that combine images, you should ensure that they are in the same colorspace. For example, they should all be CMYK, or all sRGB, or whatever. It get messy if one is CMYK and the other is sRGB. (This is like adding physical dimensions: if I travel one mile, then travel one kilometer, what total distance have I travelled? The answer is not two miles or two kilometers.) You are using the circle image only for its alpha, so I this may not create a problem for you.

If yor image is encoded as CMYK, I suggest you don't use named colours such as "white". This is because "white" is really shorthand for "#ffffff", which is the CMYK encoding for black.

Your result has a gray area. I don't see where in your code this has come from.

**Reply by neekfenwick** [reply]
Thanks @snibgo!  That Simultaneous Contrast concept is interesting.  I agree it makes sense for all images to be the same colorspace during this kind of combination.  I tried loading the images into GIMP and picking colours with the picker tool but they gave different RGB values for the same pixel between the input and output images, but never mind, I'm sure something was wrong as you pointed out.

tl/dr I think I've found a way forward, by completely bypassing colorspace transformations, PNG format, COMPOSITE_COPYOPACITY, alpha channels, and layer flattening, unfortunately it results in a black background around the cropped image, which I'm still trying to work through.

```
    $white = 'cmyk(0%,0%,0%,0%)';
    $black = 'cmyk(0%,0%,0%,100%)';

    // log_colorinfo('crop_circle before', $img);
    // Set to png for alpha channel operations
    $img->setBackgroundColor(new ImagickPixel($black));

    $circle = new Imagick();
    $circle->newImage($width, $height, 'none');
    // Setting circle colorspace to match that of main image
    $circle->setImageColorspace(Imagick::COLORSPACE_CMYK);
    // $img->profileImage('icc', file_get_contents('USWebCoatedSWOP.icc'));

    $circle->setimagematte(true);

    $draw = new ImagickDraw();
    $draw->setfillcolor(new ImagickPixel($white));
    $draw->circle($width/2, $height/2, $width/2, $height\2);
    $circle->drawimage($draw);

    $circle->compositeimage($img, Imagick::COMPOSITE_IN, 0, 0);
    $img->clear();
    $img->destroy();
```

Note that I've switched from `$img->compositeImage($circle` to `$circle->compositeImage($img`.

So a simple composite 'In' operation will take the source `$img` (the glasses of wine) and copy each pixel onto a destination image `$circle` where that image is white.  The problem here is that I actually need the background of the final image to be white (a requirement of the final printer software), so it looks lovely but the use of white as the "bits to copy" means I need a different solution.

I found this line at https://imagemagick.org/script/color-management.php : "A few formats support CMYK. For example JPG does, but PNG does not."  I am converting my JPG CMYK source image to 'png' format so it may support transparency and perform the composite COPYOPACITY overlay trick.  If the conversion to PNG results in the CMYK colorspace no longer being supported, maybe the colorspace is being converted in some weird way that screws things up, but I have no idea what's going on.

The Color Converter utility page https://imagemagick.org/contrib/color-converter.php is also interesting, it appears the parameters to the "cmyk()" function are actually Red, Green, Blue and not Cyan, Magenta, Yellow, Black as I had presumed (https://imagemagick.org/script/color.php doesn't say, annoyingly).  It also seems to imply that 'white' is valid in the context of a CMYK colour but I can't be sure.

After reading https://legacy.imagemagick.org/Usage/compose/#mask a few times, I've managed to get *exactly* what I want using the command line tools and a third 'mask' image:

```
# Make a transparent background, white circle mask image in RGB
convert -size 500x500 -colorspace RGB xc:none \
    -stroke white -strokewidth 1 -fill white \
    -draw "circle 250,250 250,10" circle_rgbmask.png

# Make an all-white image to use as the 'background' of the output
convert -size 500x500 -colorspace CMYK -profile USWebCoatedSWOP.icc xc:white -fill white -draw "rectangle 0,0 500,500" white_bg.jpg

# Yay correct result!
composite 3\ Glasses\ Of\ Wine-500.jpg white_bg.jpg circle_rgbmask.png -compose Src composed.jpg
```

So I create a pure white image just to get the final white background, and mask the wine image using a specially prepared white-circle-on-transparent PNG.

For whatever reason, the Eye of Gnome viewer (double clicking an image in Nautilus) shows the output image's colours inverted, but imagemagick's own `display` command line viewer shows the colours correctly.  I'll use `display` to verify results from now on.

However, the PHP Imagick bindings' `compositeImage()` function does not seem to support a 'mask' image (the `circle_rgbmask.png` in the example above) :(

I'm going to back to the PHP and try to replicate what I've achieved on the command line.
